pkg/util/winutil/scmmonitor.go

```diff
@@ -0,0 +1,203 @@
+
+// GetRefreshCount returns the number of times we've actually queried
+// the SCM database.  used for logging stats.
+func (scm *SCMMonitor) GetRefreshCount() uint64 {
+	return scm.serviceRefreshes
+}
+func (scm *SCMMonitor) refreshCache() error {
+
+	// EnumServiceStatusEx requires only SC_MANAGER_ENUM_SERVICE.  Switch to
+	// new library to use least privilege
+	h, err := windows.OpenSCManager(nil, nil, windows.SC_MANAGER_ENUMERATE_SERVICE)
+	if err != nil {
+		log.Warnf("Failed to connect to scm %v", err)
+		return fmt.Errorf("Failed to open SCM %v", err)
+	}
+	m := &mgr.Mgr{Handle: h}
+	defer m.Disconnect()
+
+	var bytesNeeded, servicesReturned uint32
+	var buf []byte
+	for {
+		var p *byte
+		if len(buf) > 0 {
+			p = &buf[0]
+		}
+		err = windows.EnumServicesStatusEx(m.Handle, windows.SC_ENUM_PROCESS_INFO,
+			windows.SERVICE_WIN32, windows.SERVICE_STATE_ALL,
+			p, uint32(len(buf)), &bytesNeeded, &servicesReturned, nil, nil)
+		if err == nil {
+			break
+		}
+		if err != windows.ERROR_MORE_DATA {
+			return fmt.Errorf("Failed to enum services %v", err)
+		}
+		if bytesNeeded <= uint32(len(buf)) {
+			return err
+		}
+		buf = make([]byte, bytesNeeded)
+	}
+	if servicesReturned == 0 {
+		return nil
+	}
+
+	//var services []windows.ENUM_SERVICE_STATUS_PROCESS
+	//services := (*[1<<29 - 1]windows.ENUM_SERVICE_STATUS_PROCESS)(unsafe.Pointer(&buf[0]))[:servicesReturned]
+	services := unsafe.Slice((*windows.ENUM_SERVICE_STATUS_PROCESS)(unsafe.Pointer(&buf[0])), servicesReturned)
+
+	//hdr.Data = unsafe.Pointer(&buf[0])
+	//hdr.Len = int(servicesReturned)
+	//hdr.Cap = int(servicesReturned)
+
+	newmap := make(map[uint64]*Service)
+	for idx, svc := range services {
+		var thissvc *Service
+		var ok bool
+		if thissvc, ok = newmap[uint64(svc.ServiceStatusProcess.ProcessId)]; !ok {
+			thissvc = &Service{}
+			newmap[uint64(svc.ServiceStatusProcess.ProcessId)] = thissvc
+		}
+		thissvc.essp = append(thissvc.essp, &services[idx])
+	}
+	var current windows.Filetime
+	windows.GetSystemTimeAsFileTime(&current)
+	scm.lastMapTime = uint64(current.Nanoseconds())
+	scm.pidToService = newmap
+	scm.serviceRefreshes++
+	return nil
+}
+
+func (s *Service) toServiceInfo() *ServiceInfo {
+	var si ServiceInfo
+	for _, inf := range s.essp {
+		si.DisplayName = append(si.DisplayName, windows.UTF16PtrToString(inf.DisplayName))
+		si.ServiceName = append(si.ServiceName, windows.UTF16PtrToString(inf.ServiceName))
+	}
+	return &si
+
+}
+
+func (scm *SCMMonitor) getServiceFromCache(pid, pidstart uint64) (*ServiceInfo, error) {
+	var err error
+	if val, ok := scm.pidToService[pid]; ok {
+		// it was in there.  see if it's the right one
+		if val.startTime == 0 {
+			// we don't prepopulate the cache with pid start times
+			val.startTime, err = pGetProcessStartTimeAsNs(pid)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if val.startTime == pidstart {
+			// it's the same one.
+			return val.toServiceInfo(), nil
```

Comment:
Should you cache results in place instead of keep converting raw bytes to strings for the same pid?

Reply:
actually, it's done this way intentionally.  The entire cache is a single, large buffer created every time we used the cache.  If we return it in-place, then the caller could hold on to a single element, and cause the entire buffer to not be garbage collected.
This way, the caller retains ownership of the strings, but we retain ownership of the large buffer.